useEffect reference (#431)
reactjs/pl.react.dev: src/content/reference/react/useEffect.md

* Jeśli twój efekt nie został wywołany przez interakcję (np. kliknięcie), React zazwyczaj pozwoli przeglądarce **najpierw odświeżyć ekran przed uruchomieniem twojego efektu.** Jeśli efekt ten wykonuje jakieś operacje związane z wyświetlaniem (np. ustawianie pozycji dymka (ang. *tooltip*)) i opóźnienie jest zauważalne (np. występuje migotanie), zastąp `useEffect` przez [`useLayoutEffect`.](/reference/react/useLayoutEffect)

* Nawet jeśli twój efekt został wywołany przez interakcję (np. kliknięcie), **przeglądarka może odświeżyć ekran przed przetworzeniem aktualizacji stanu wewnątrz twojego efektu.** Zazwyczaj jest to pożądane zachowanie. Niemniej jednak, jeśli chcesz zablokować przeglądarkę przed odświeżaniem ekranu, musisz zastąpić `useEffect` przez [`useLayoutEffect`.](/reference/react/useLayoutEffect)

* Efekty uruchamiane są **tylko po stronie klienta.** Nie są uruchamiane podczas renderowania po stronie serwera.

---

## Sposób użycia {/*usage*/}

### Łączenie z zewnętrznym systemem {/*connecting-to-an-external-system*/}

> Niektóre komponenty, gdy są wyświetlane na stronie, muszą pozostać połączone z siecią, pewnym interfejsem przeglądarki lub zewnętrzną biblioteką. Systemy te nie są kontrolowane przez Reacta, dlatego nazywane są *zewnętrznymi.*

Reviewer: Za dużo wtrąceń, przez co zdanie jest mocno poszatkowane i ciężko je zrozumieć.

```suggestion
Niektóre komponenty muszą pozostać połączone z siecią, pewnym interfejsem przeglądarki lub zewnętrzną biblioteką podczas wyświetlania ich na ekranie. Systemy te nie są kontrolowane przez Reacta, dlatego nazywane są *zewnętrznymi.*
```

}
```

> Aby użyć `useEffect`, musisz przekazać dwie argumenty:

Reviewer: ```suggestion
Aby użyć `useEffect`, musisz przekazać dwa argumenty:
```

> 1. *Funkcję konfiguracyjną* z <CodeStep step={1}>kodem konfiguracyjnym</CodeStep>, który łączy się z tym systemem.

Reviewer: ```suggestion
1. *Funkcję konfiguracyjną* z <CodeStep step={1}>kodem konfigurującym</CodeStep>, który łączy się z tym systemem.
```

   - Funkcja ta powinna zwracać *funkcję czyszczącą* z <CodeStep step={2}>kodem czyszczącym</CodeStep>, który rozłącza się z tym systemem.

> 2. <CodeStep step={3}>Tablicę zależności</CodeStep>, zawierającą każdą wartość używaną wewnątrz tych funkcji w twoim komponencie.

Reviewer: ```suggestion
2. <CodeStep step={3}>Tablicę zależności</CodeStep>, zawierającą każdą wartość z twojego komponentu używaną wewnątrz tych funkcji.
```

**React wywołuje twoje funkcje konfiguracyjną i czyszczącą wtedy, gdy jest to konieczne, co może się zdarzyć wielokrotnie:**

> 1. Twój <CodeStep step={1}>kod konfiguracyjny</CodeStep> jest wykonywany, gdy twój komponent jest dodawany do strony *(montowany)*.

Reviewer: ```suggestion
1. Twój <CodeStep step={1}>kod konfigurujący</CodeStep> jest wykonywany, gdy komponent jest dodawany do strony *(montowany)*.
```

2. Po każdym renderowaniu twojego komponentu, w którym <CodeStep step={3}>zależności</CodeStep> uległy zmianie:
   - Najpierw jest wykonywany twój <CodeStep step={2}>kod czyszczący</CodeStep> z poprzednimi właściwościami i stanem.

>    - Następnie jest wykonywany twój <CodeStep step={1}>kod konfiguracyjny</CodeStep> z nowymi właściwościami i stanem.

Reviewer: ```suggestion
   - Następnie jest wykonywany twój <CodeStep step={1}>kod konfigurujący</CodeStep> z nowymi właściwościami i stanem.
```

> 3. Twój <CodeStep step={2}>kod czyszczący</CodeStep> jest wykonywany jeszcze raz po usunięciu *(odmontowaniu)* twojego komponentu ze strony.

Reviewer: ```suggestion
3. Twój <CodeStep step={2}>kod czyszczący</CodeStep> jest wykonywany jeszcze raz po usunięciu *(odmontowaniu)* komponentu ze strony.
```

> **Przyjrzyjmy się tej sekwencji z przykładu powyżej.**

Reviewer: ```suggestion
**Przyjrzyjmy się tej sekwencji dla przykładu powyżej.**
```

> Kiedy komponent `ChatRoom` zostanie dodany do strony, połączy się z pokojem czatu przy użyciu początkowych `serverUrl` i `roomId`. Jeśli którakolwiek z zależności `serverUrl` lub `roomId` zmieni się w wyniku przerenderowania (np. jeśli użytkownik wybierze inny pokój czatu z rozwijanej listy), twój efekt *rozłączy się z poprzednim pokojem i połączy się z następnym.* Kiedy komponent `ChatRoom` zostanie usunięty ze strony, twój efekt rozłączy się ostatni raz.

Reviewer: ```suggestion
Kiedy komponent `ChatRoom` zostanie dodany do strony, połączy się z pokojem czatu przy użyciu początkowych `serverUrl` i `roomId`. Jeśli którakolwiek z zależności - `serverUrl` lub `roomId` - zmieni się w wyniku przerenderowania (np. jeśli użytkownik wybierze inny pokój czatu z rozwijanej listy), twój efekt *rozłączy się z poprzednim pokojem i połączy się z następnym.* Kiedy komponent `ChatRoom` zostanie usunięty ze strony, twój efekt rozłączy się ostatni raz.
```

> **Aby [pomóc w wykrywaniu błędów,](/learn/synchronizing-with-effects#step-3-add-cleanup-if-needed) w trybie developerskim React wykonuje dodatkowo funkcję <CodeStep step={1}>setup</CodeStep> oraz <CodeStep step={2}>cleanup</CodeStep> przed właściwym wywołaniem <CodeStep step={1}>setup</CodeStep>.** Jest to test obciążeniowy, który sprawdza, czy logika twojego efektu jest poprawnie zaimplementowana. Jeśli to spowoduje widoczne problemy, oznacza to, że brakuje pewnej logiki w funkcji czyszczącej. Funkcja ta powinna zatrzymać lub cofnąć wszystko, co zrobiła funkcja konfiguracyjna. Ogólnie rzecz biorąc, użytkownik nie powinien być w stanie rozróżnić między jednorazowym wywołaniem konfiguracji (jak na produkcji), a sekwencją *konfiguracja* → *czyszczenie* → *konfiguracja* (jak w trybie developerskim). [Zobacz najczęstsze rozwiązania.](/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development)

Reviewer: ```suggestion
**Aby [pomóc w wykrywaniu błędów,](/learn/synchronizing-with-effects#step-3-add-cleanup-if-needed) w trybie developerskim React wykonuje dodatkowo funkcję <CodeStep step={1}>konfigurującą</CodeStep> oraz <CodeStep step={2}>czyszczącą</CodeStep> przed właściwym, docelowym wywołaniem <CodeStep step={1}>konfigurującej</CodeStep>.** Jest to test obciążeniowy, który sprawdza, czy logika twojego efektu jest poprawnie zaimplementowana. Jeśli to spowoduje widoczne problemy, oznacza to, że brakuje pewnej logiki w funkcji czyszczącej. Funkcja ta powinna zatrzymać lub cofnąć wszystko, co zrobiła funkcja konfiguracyjna. Ogólnie rzecz biorąc, użytkownik nie powinien być w stanie rozróżnić między jednorazowym wywołaniem konfiguracji (jak na produkcji), a sekwencją *konfiguracja* → *czyszczenie* → *konfiguracja* (jak w trybie developerskim). [Zobacz najczęstsze rozwiązania.](/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development)
```

> **Postaraj się pisać każdy efekt jako niezależny proces** i **skup się na pojedynczym cyklu konfiguracji i czyszczenia w danym momencie.** Nie ma znaczenia, czy komponent jest montowany, aktualizowany czy odmontowywany. Jeśli logika czyszczenia poprawnie odwzorowuje logikę konfiguracji, twój efekt jest odporny na uruchamianie konfiguracji i czyszczenia tak często, jak to konieczne.

Reviewer: ```suggestion
**Postaraj się pisać każdy efekt jako niezależny proces** i **skup się na pojedynczym cyklu konfigurowania i czyszczenia w danym momencie.** Nie ma znaczenia, czy komponent jest montowany, aktualizowany czy odmontowywany. Jeśli logika czyszczenia poprawnie odwzorowuje logikę konfigurowania, twój efekt jest odporny na uruchamianie konfigurowania i czyszczenia tak często, jak to konieczne.
```

* Timer zarządzany za pomocą <CodeStep step={1}>[`setInterval()`](https://developer.mozilla.org/en-US/docs/Web/API/setInterval)</CodeStep> i <CodeStep step={2}>[`clearInterval()`](https://developer.mozilla.org/en-US/docs/Web/API/clearInterval)</CodeStep>.
* Subskrypcja zdarzeń za pomocą <CodeStep step={1}>[`window.addEventListener()`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener)</CodeStep> i <CodeStep step={2}>[`window.removeEventListener()`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener)</CodeStep>.
* Zewnętrzna biblioteka animacji z API w stylu <CodeStep step={1}>`animation.start()`</CodeStep> i <CodeStep step={2}>`animation.reset()`</CodeStep>.

**Jeśli nie łączysz się z żadnym zewnętrznym systemem, [prawdopodobnie nie potrzebujesz efektu.](/learn/you-might-not-need-an-effect)**

</Note>

<Recipes titleText="Przykłady łączenia się z systemem zewnętrznym" titleId="examples-connecting">

#### Łączenie się z serwerem czatu {/*connecting-to-a-chat-server*/}

> W tym przykładzie komponent `ChatRoom` wykorzystuje efekt do utrzymania połączenia z systemem zewnętrznym zdefiniowanym w pliku `chat.js`. Naciśnij "Otwórz czat", aby pojawił się komponent `ChatRoom`. Ta piaskownica działa w trybie developerskim, więc ma miejsce dodatkowy cykl łączenia i rozłączania, tak jak jest to [wyjaśnione tutaj](/learn/synchronizing-with-effects#step-3-add-cleanup-if-needed). Spróbuj zmieniać `roomId` i `serverUrl` za pomocą rozwijanej listy i pola tekstowego, a zobaczysz, jak efekt ponownie łączy się z czatem. Naciśnij "Zamknij czat", aby zobaczyć jak efekt kończy połączenie ostatni raz.

Reviewer: ```suggestion
W tym przykładzie komponent `ChatRoom` wykorzystuje efekt do utrzymania połączenia z systemem zewnętrznym zdefiniowanym w pliku `chat.js`. Naciśnij "Otwórz czat", aby pojawił się komponent `ChatRoom`. Ta piaskownica działa w trybie developerskim, więc ma miejsce dodatkowy cykl łączenia i rozłączania, tak jak jest to [wyjaśnione tutaj](/learn/synchronizing-with-effects#step-3-add-cleanup-if-needed). Spróbuj zmieniać `roomId` i `serverUrl` za pomocą rozwijanej listy i pola tekstowego, a zobaczysz, jak efekt ponownie łączy się z czatem. Naciśnij "Zamknij czat", aby zobaczyć, jak efekt kończy połączenie po raz ostatni.
```

  return (
    <>
      <label>
        URL serwera:{' '}
        <input
          value={serverUrl}
          onChange={e => setServerUrl(e.target.value)}
        />
      </label>
      <h1>Witaj w pokoju: {roomId}!</h1>
    </>
  );
}

export default function App() {

>   const [roomId, setRoomId] = useState('ogólny');

Reviewer: ```suggestion
  const [roomId, setRoomId] = useState('general');
```

  const [show, setShow] = useState(false);
  return (
    <>
      <label>
        Wybierz pokój czatu:{' '}
        <select
          value={roomId}
          onChange={e => setRoomId(e.target.value)}
        >

>           <option value="ogólny">ogólny</option>

Reviewer: ```suggestion
          <option value="general">ogólny</option>
```

>           <option value="podróże">podróże</option>

Reviewer: ```suggestion
          <option value="travel">podróże</option>
```

>           <option value="muzyka">muzyka</option>

Reviewer: ```suggestion
          <option value="music">muzyka</option>
```

<Solution />

#### Wywoływanie animacji {/*triggering-an-animation*/}

> W tym przykładzie, systemem zewnętrznym jest biblioteka animacji w pliku `animation.js`. Udostępnia ona klasę javascriptową o nazwie `FadeInAnimation`, która przyjmuje jako argument węzeł DOM i udostępnia metody `start()` oraz `stop()` do sterowania animacją. Ten komponent [używa referencji (ang. *ref*)](/learn/manipulating-the-dom-with-refs), aby mieć dostęp do zasadniczego węzła DOM. Gdy komponent staje się widoczny, efekt odczytuje ten węzeł z referencji i automatycznie uruchamia dla niego animację.

Reviewer: ```suggestion
W tym przykładzie, systemem zewnętrznym jest biblioteka animacji w pliku `animation.js`. Udostępnia ona klasę javascriptową o nazwie `FadeInAnimation`, która przyjmuje jako argument węzeł DOM i udostępnia metody `start()` oraz `stop()` do sterowania animacją. Ten komponent [używa referencji (ang. *ref*)](/learn/manipulating-the-dom-with-refs), aby mieć dostęp do bazowego węzła DOM. Gdy komponent staje się widoczny, efekt odczytuje ten węzeł z referencji i automatycznie uruchamia dla niego animację.
```

<Solution />

#### Śledzenie widoczności elementu {/*tracking-element-visibility*/}

> W tym przykładzie, systemem zewnętrznym ponownie jest drzewo DOM w przeglądarce. Komponent `App` wyświetla długą listę, następnie komponent `Box`, a potem kolejną długą listę. Przewiń w dół tej listy. Zauważ, że gdy komponent `Box` pojawia się cały w widocznym obszarze przeglądarki, kolor tła zmienia się na czarny. Aby to zaimplementować, komponent `Box` używa efektu do sterowania [`IntersectionObserver`](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API). To API przeglądarki powiadamia, kiedy element DOM jest widoczny.

Reviewer: ```suggestion
W tym przykładzie, systemem zewnętrznym ponownie jest drzewo DOM w przeglądarce. Komponent `App` wyświetla długą listę, następnie komponent `Box`, a potem kolejną długą listę. Przewiń w dół tej listy. Zauważ, że gdy komponent `Box` pojawia się cały w widocznym obszarze przeglądarki, kolor tła zmienia się na czarny. Aby to zaimplementować, komponent `Box` używa efektu do sterowania [`IntersectionObserver`](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API). To API przeglądarki powiadamia naszą funkcję, kiedy element DOM jest widoczny.
```

  return (
    <>
      <label>
        URL serwera:{' '}
        <input
          value={serverUrl}
          onChange={e => setServerUrl(e.target.value)}
        />
      </label>
      <h1>Witaj w pokoju: {roomId}!</h1>
    </>
  );
}

export default function App() {

>   const [roomId, setRoomId] = useState('ogólny');

Reviewer: ```suggestion
  const [roomId, setRoomId] = useState('general');
```

  const [show, setShow] = useState(false);
  return (
    <>
      <label>
        Wybierz pokój czatu:{' '}
        <select
          value={roomId}
          onChange={e => setRoomId(e.target.value)}
        >

>           <option value="ogólny">ogólny</option>

Reviewer: ```suggestion
          <option value="general">ogólny</option>
```

>           <option value="podróże">podróże</option>

Reviewer: ```suggestion
          <option value="travel">podróże</option>
```

>           <option value="muzyka">muzyka</option>

Reviewer: ```suggestion
          <option value="music">muzyka</option>
```

<Solution />

#### Własny hook `useWindowListener` {/*custom-usewindowlistener-hook*/}

> Ten przykład jest identyczny jak [jeden z wcześniejszych przykładów,](#examples-connecting) ale logika została wyodrębniona do własnego hooka.

Reviewer: ```suggestion
Ten przykład jest identyczny z [jednym z wcześniejszych,](#examples-connecting) ale logika została wyodrębniona do własnego hooka.
```

<Solution />

#### Własny hook `useIntersectionObserver` {/*custom-useintersectionobserver-hook*/}

> Ten przykład jest identyczny jak [jeden z wcześniejszych przykładów,](#examples-connecting) ale logika została częściowo wyodrębniona do własnego hooka.

Reviewer: ```suggestion
Ten przykład jest identyczny z [jednym z wcześniejszych,](#examples-connecting) ale logika została częściowo wyodrębniona do własnego hooka.
```

  // ...
```

> Zwróć uwagę na zmienną `ignore`, która jest inicjowana jako `false` i ustawiana na `true` podczas czyszczenia. Zapewnia to, że [twój kod nie będzie podatny na tzw. "hazardy" (ang. *race conditions*):](https://maxrozen.com/race-conditions-fetching-data-react-with-useeffect) odpowiedzi z sieci mogą przychodzić w innej kolejności, niż zostały wysłane żądania.

Reviewer: ```suggestion
Zwróć uwagę na zmienną `ignore`, która jest inicjowana jako `false` i ustawiana na `true` podczas czyszczenia. Zapewnia to, że [twój kod nie będzie podatny na tzw. "wyścigi" (ang. *race conditions*):](https://maxrozen.com/race-conditions-fetching-data-react-with-useeffect) odpowiedzi z sieci mogą przychodzić w innej kolejności, niż zostały wysłane żądania.
```

      <select value={person} onChange={e => {
        setPerson(e.target.value);
      }}>
        <option value="Alicja">Alicja</option>
        <option value="Barbara">Barbara</option>
        <option value="Tadeusz">Tadeusz</option>
      </select>
      <hr />
      <p><i>{bio ?? 'Ładowanie...'}</i></p>
    </>
  );
}
```

```js api.js hidden
export async function fetchBio(person) {
  const delay = person === 'Barbara' ? 2000 : 200;
  return new Promise(resolve => {
    setTimeout(() => {
      resolve( person + ' - biografia.');
    }, delay);
  })
}
```

</Sandpack>

> Możesz to również przepisać, używając składni [`async` / `await`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function), ale nadal musisz napisać funkcję czyszczącą:

Reviewer: ```suggestion
Możesz również przepisać ten kod używając składni [`async` / `await`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function), ale nadal musisz napisać funkcję czyszczącą:
```

- **Efekty nie są uruchamiane na serwerze.** Oznacza to, że początkowy HTML renderowany po stronie serwera będzie zawierał jedynie stan ładowania bez danych. Komputer klienta musiałby pobrać cały kod JavaScript i wyrenderować aplikację, tylko po to, by odkryć, że teraz musi pobrać dane. To nie jest zbyt wydajne podejście.
- **Bezpośrednie pobieranie danych w efektach sprzyja tworzeniu "kaskad żądań sieciowych" (ang. *network waterfall*).** Renderujesz komponent rodzica, on pobiera pewne dane, renderuje komponenty potomne, a następnie one zaczynają pobierać swoje dane. Jeśli sieć nie jest zbyt szybka, takie podejście jest to znacznie wolniejsze niż równoczesne pobieranie wszystkich danych.
- **Pobieranie bezpośrednio w efektach zazwyczaj oznacza brak wstępnego wczytywania (ang. *preload*) i buforowania danych (ang. *cache*).** Na przykład, jeśli komponent jest odmontowywany, a następnie ponownie montowany, będzie trzeba ponownie pobrać dane.

> - **Nie jest to zbyt ergonomiczne.** Pisanie wywołania `fetch` w taki sposób, aby uniknąć błędów takich jak [hazardy](https://maxrozen.com/race-conditions-fetching-data-react-with-useeffect), wymaga dość dużej ilości kodu.

Reviewer: ```suggestion
- **Nie jest to zbyt ergonomiczne.** Pisanie wywołania `fetch` w taki sposób, aby uniknąć błędów takich jak [wyścigi](https://maxrozen.com/race-conditions-fetching-data-react-with-useeffect), wymaga dość dużej ilości kodu.
```

Te wady nie dotyczą tylko Reacta. Występują one przy pobieraniu danych podczas montowania komponentu przy użyciu dowolnej biblioteki. Podobnie jak w przypadku routingu, poprawne wykonywanie pobierania danych nie jest proste, dlatego polecamy następujące podejścia:

- **Jeśli używasz [frameworka](/learn/start-a-new-react-project#production-grade-react-frameworks), wykorzystaj jego wbudowany mechanizm pobierania danych.** Współczesne frameworki reactowe mają zintegrowane mechanizmy pobierania danych, które są wydajne i rozwiązują powyższe problemy.

> - **W przeciwnym razie, rozważ użycie lub zbudowanie pamięci podręcznej po stronie klienta.** Popularne rozwiązania open source obejmują [React Query](https://react-query.tanstack.com/), [useSWR](https://swr.vercel.app/) oraz [React Router 6.4+.](https://beta.reactrouter.com/en/main/start/overview) Możesz także zbudować swoje własne rozwiązanie, w którym byłyby użyte efekty, ale także byłaby logika do unikania zduplikowanych zapytań, buforowania odpowiedzi i unikania kaskad żądań sieciowych (poprzez wstępne wczytywanie danych lub przeniesienie wymagań dot. danych do ścieżek).

Reviewer: ```suggestion
- **W przeciwnym razie, rozważ użycie lub zbudowanie pamięci podręcznej po stronie klienta.** Popularne rozwiązania open source obejmują [React Query](https://react-query.tanstack.com/), [useSWR](https://swr.vercel.app/) oraz [React Router 6.4+.](https://beta.reactrouter.com/en/main/start/overview) Możesz także zbudować swoje własne rozwiązanie, w którym byłyby użyte efekty, ale także zawarte byłyby: logika do unikania zduplikowanych zapytań, buforowania odpowiedzi i unikania kaskad żądań sieciowych (poprzez wstępne wczytywanie danych lub przeniesienie wymagań dot. danych do ścieżek).
```

    const connection = createConnection(serverUrl, roomId);
    connection.connect();
    return () => connection.disconnect();
  }, []); // 🔴 Reactowy hook useEffect ma brakujące zależności: 'roomId' i 'serverUrl'
  // ...
}
```

> **Aby usunąć zależność, musisz "udowodnić" linterowi, że to *nie musi być* zależność.** Na przykład, możesz przenieść `serverUrl` poza swój komponent i tym samym udowodnić, że nie jest wartość reaktywna i nie zmieni się podczas ponownego renderowania:

Reviewer: ```suggestion
**Aby usunąć zależność, musisz "udowodnić" linterowi, że to *nie musi być* zależność.** Na przykład, możesz przenieść `serverUrl` poza swój komponent i tym samym udowodnić, że nie jest wartością reaktywną i nie zmieni się podczas ponownego renderowania:
```

function ChatRoom() {
  useEffect(() => {
    const connection = createConnection(serverUrl, roomId);
    connection.connect();
    return () => connection.disconnect();
  }, []); // ✅ Wszystkie zależności zadeklarowane
  // ...
}
```

[Efekt z pustymi zależnościami](/learn/lifecycle-of-reactive-effects#what-an-effect-with-empty-dependencies-means) nie zostanie ponownie uruchomiony, gdy zmieni się którakolwiek właściwość lub stan twojego komponentu.

<Pitfall>

W już istniejącym kodzie, możesz mieć pewne efekty, które uciszają lintera w taki sposób:

```js {3-4}
useEffect(() => {
  // ...
  // 🔴 Unikaj uciszania lintera w taki sposób:
  // eslint-ignore-next-line react-hooks/exhaustive-deps
}, []);
```

**Kiedy zależności nie pasują do kodu, istnieje duże ryzyko wprowadzenia błędów.** Uciszając lintera, "oszukujesz" Reacta co do wartości, od których zależy twój efekt. [Zamiast tego, udowodnij, że są one zbędne.](/learn/removing-effect-dependencies#removing-unnecessary-dependencies)

</Pitfall>

<Recipes titleText="Przykłady przekazywania reaktywnych zależności" titleId="examples-dependencies">

#### Przekazywanie tablicy zależności {/*passing-a-dependency-array*/}

> Jeśli określisz zależności, twój efekt zostanie uruchomiony **po początkowym renderowaniu _oraz_ po ponownym renderowaniu z zmienionymi zależnościami.**

Reviewer: ```suggestion
Jeśli określisz zależności, twój efekt zostanie uruchomiony **po początkowym renderowaniu _oraz_ po ponownym renderowaniu ze zmienionymi zależnościami.**
```

```js {3}
useEffect(() => {
  // ...

> }, [a, b]); // Zostanie uruchomiony ponownie, jeśli a lub b są różne

Reviewer: ```suggestion
}, [a, b]); // Zostanie uruchomiony ponownie, jeśli a lub b ulegną zmianie
```

export default function App() {
  const [show, setShow] = useState(false);

>   const [roomId, setRoomId] = useState('ogólny');

Reviewer: ```suggestion
  const [roomId, setRoomId] = useState('general');
```

  return (
    <>
      <label>
        Wybierz pokój czatu:{' '}
        <select
          value={roomId}
          onChange={e => setRoomId(e.target.value)}
        >

>           <option value="ogólny">ogólny</option>

Reviewer: ```suggestion
          <option value="general">ogólny</option>
```

>           <option value="podróże">podróże</option>

Reviewer: ```suggestion
          <option value="travel">podróże</option>
```

>           <option value="muzyka">muzyka</option>

Reviewer: ```suggestion
          <option value="music">muzyka</option>
```

import { createConnection } from './chat.js';

const serverUrl = 'https://localhost:1234';

> const roomId = 'muzyka';

Reviewer: ```suggestion
const roomId = 'music';
```

export default function App() {
  const [show, setShow] = useState(false);

>   const [roomId, setRoomId] = useState('ogólny');

Reviewer: ```suggestion
  const [roomId, setRoomId] = useState('general');
```

  return (
    <>
      <label>
        Wybierz pokój czatu:{' '}
        <select
          value={roomId}
          onChange={e => setRoomId(e.target.value)}
        >

>           <option value="ogólny">ogólny</option>

Reviewer: ```suggestion
          <option value="general">ogólny</option>
```

>           <option value="podróże">podróże</option>

Reviewer: ```suggestion
          <option value="travel">podróże</option>
```

>           <option value="muzyka">muzyka</option>

Reviewer: ```suggestion
          <option value="music">muzyka</option>
```

  return (
    <>
      <h1>Witaj w pokoju: {roomId}</h1>
      <input value={message} onChange={e => setMessage(e.target.value)} />
    </>
  );
}

export default function App() {

>   const [roomId, setRoomId] = useState('ogólny');

Reviewer: ```suggestion
  const [roomId, setRoomId] = useState('general');
```

  return (
    <>
      <label>
        Wybierz pokój czatu:{' '}
        <select
          value={roomId}
          onChange={e => setRoomId(e.target.value)}
        >

>           <option value="ogólny">ogólny</option>

Reviewer: ```suggestion
          <option value="general">ogólny</option>
```

>           <option value="podróże">podróże</option>

Reviewer: ```suggestion
          <option value="travel">podróże</option>
```

>           <option value="muzyka">muzyka</option>

Reviewer: ```suggestion
          <option value="music">muzyka</option>
```

- [Odczytywanie najnowszych właściwości i stanu z efektu](#reading-the-latest-props-and-state-from-an-effect)

Ostatecznym rozwiązaniem (jeśli powyższe metody nie pomogły) jest opakowanie tworzenia obiektu w [`useMemo`](/reference/react/useMemo#memoizing-a-dependency-of-another-hook) lub [`useCallback`](/reference/react/useCallback#preventing-an-effect-from-firing-too-often) (dla funkcji).

---

### Mój efekt wpada w nieskończoną pętlę {/*my-effect-keeps-re-running-in-an-infinite-cycle*/}

Jeżeli twój efekt wpada w nieskończoną pętlę, te dwa warunki muszą być spełnione:

- Twój efekt aktualizuje jakiś stan.
- Ten stan prowadzi do przerenderowania, co powoduje zmiany w zależnościach efektu.

Zanim ruszysz do naprawiania problemu, zastanów się, czy twój efekt nie łączy się z jakimś zewnętrznym systemem (takim jak drzewo DOM, sieć, widżet itp.). Dlaczego twój efekt musi ustawiać stan? Czy synchronizuje się on z tym zewnętrznym systemem? Czy próbujesz za jego pomocą zarządzać przepływem danych w twojej aplikacji?

Jeśli nie ma żadnego zewnętrznego systemu, zastanów się, czy [całkowite usunięcie efektu](/learn/you-might-not-need-an-effect) nie uprościłoby twojej logiki.

> Jeśli rzeczywiście synchronizujesz się z jakimś zewnętrznym systemem, zastanów się, dlaczego i pod jakim warunkiem twój efekt powinien aktualizować stan. Czy zmieniło się coś, co wpływa na to jak powinien wyglądać twój komponent? Jeśli musisz śledzić jakieś dane, które nie są używane do renderowania, może bardziej odpowiednie będzie użycie [referencji](/reference/react/useRef#referencing-a-value-with-a-ref) (która nie powoduje przerenderowania). Upewnij się, że twój efekt nie aktualizuje stanu (i nie powoduje ponownych renderowań) częściej niż to konieczne.

Reviewer: ```suggestion
Jeśli rzeczywiście synchronizujesz się z jakimś zewnętrznym systemem, zastanów się, dlaczego i pod jakim warunkiem twój efekt powinien aktualizować stan. Czy zmieniło się coś, co wpływa na to, jak powinien wyglądać twój komponent? Jeśli musisz śledzić jakieś dane, które nie są używane do renderowania, może bardziej odpowiednie będzie użycie [referencji](/reference/react/useRef#referencing-a-value-with-a-ref) (która nie powoduje przerenderowania). Upewnij się, że twój efekt nie aktualizuje stanu (i nie powoduje ponownych renderowań) częściej niż to konieczne.
```

Kończąc, jeśli twój efekt aktualizuje stan w odpowiednim momencie, ale wciąż występuje zapętlenie, oznacza to, że ta aktualizacja stanu prowadzi do zmiany jednej z zależności efektu. [Przeczytaj, jak debugować zmiany w zależnościach.](/reference/react/useEffect#my-effect-runs-after-every-re-render)

---

> ### Moja funkcja czyszcząca jest uruchamiana nawet, gdy mój komponent nie jest odmontowywany {/*my-cleanup-logic-runs-even-though-my-component-didnt-unmount*/}

Reviewer: ```suggestion
### Moja funkcja czyszcząca jest uruchamiana nawet wtedy, gdy mój komponent nie jest odmontowywany {/*my-cleanup-logic-runs-even-though-my-component-didnt-unmount*/}
```

> Funkcja czyszcząca uruchamia się nie tylko podczas odmontowywania, ale również przed każdym przerenderowaniem ze zmienionymi zależnościami. Dodatkowo, w trybie deweloperskim, React [uruchamia funkcję konfiguracyjną oraz czyszczącą dodatkowy raz, tuż po zamontowaniu komponentu.](#my-effect-runs-twice-when-the-component-mounts)

Reviewer: ```suggestion
Funkcja czyszcząca uruchamia się nie tylko podczas odmontowywania, ale również przed każdym przerenderowaniem ze zmienionymi zależnościami. Dodatkowo, w trybie deweloperskim, React [uruchamia funkcję konfigurującą oraz czyszczącą dodatkowy raz, tuż po zamontowaniu komponentu.](#my-effect-runs-twice-when-the-component-mounts)
```

> Jeśli masz kod czyszczący bez odpowiadającego mu kodu konfigurującego, zazwyczaj oznacza to, że kod ten może powodować problemy:

Reviewer: ```suggestion
Jeśli masz kod czyszczący bez odpowiadającego mu kodu konfigurującego, zazwyczaj to on jest przyczyną problemów:
```

```js {2-5}
useEffect(() => {

>   // 🔴 Unikaj: Logika czyszczenia bez odpowiadającej jej logiki konfiguracyjnej.

Reviewer: ```suggestion
  // 🔴 Unikaj: Logika czyszczenia bez odpowiadającej jej logiki konfigurującej.
```

  return () => {
    doSomething();
  };
}, []);
```

> Twoja logika czyszcząca powinna być "symetryczna" względem logiki konfiguracyjnej i powinna zatrzymać lub cofać to, co zrobiła funkcja konfiguracyjna:

Reviewer: ```suggestion
Twoja logika czyszcząca powinna być "symetryczna" względem logiki konfigurującej i powinna zatrzymać lub wycofać to, co zrobiła funkcja konfigurująca:
```

  }, [serverUrl, roomId]);
```

[Dowiedz się, jak cykl życia efektu różni się od cyklu życia komponentu.](/learn/lifecycle-of-reactive-effects#the-lifecycle-of-an-effect)

---

### Mój efekt robi coś wizualnego i widzę migotanie przed jego uruchomieniem {/*my-effect-does-something-visual-and-i-see-a-flicker-before-it-runs*/}

> Jeśli Twój efekt musi wstrzymać przeglądarkę przed [pokazaniem zawartości na ekranie,](/learn/render-and-commit#epilogue-browser-paint) zastąp `useEffect` z pomocą [`useLayoutEffect`](/reference/react/useLayoutEffect). Pamiętaj, że **to nie będzie konieczne w przypadku zdecydowanej większości efektów.** Będziesz tego potrzebować tylko wtedy, gdy konieczne jest uruchomienie efektu przed tym jak przeglądarka zacznie wyświetlać zawartość, na przykład do pomiaru i pozycjonowania dymka podpowiedzi zanim użytkownik go zobaczy.

Reviewer: ```suggestion
Jeśli twój efekt musi wstrzymać przeglądarkę przed [pokazaniem zawartości na ekranie,](/learn/render-and-commit#epilogue-browser-paint) zamień `useEffect` na [`useLayoutEffect`](/reference/react/useLayoutEffect). Pamiętaj, że **nie jest to konieczne w przypadku zdecydowanej większości efektów.** Będziesz tego potrzebować tylko wtedy, gdy konieczne jest uruchomienie efektu przed tym, jak przeglądarka zacznie wyświetlać zawartość, na przykład do pomiaru i pozycjonowania dymka podpowiedzi, zanim użytkownik go zobaczy.
```
